test/extensions/filters/http/common/compressor/websocket_with_compressor_integration_test.cc

```diff
@@ -0,0 +1,248 @@
+              ->set_allow_connect(true);
+          ConfigHelper::setProtocolOptions(
+              *bootstrap.mutable_static_resources()->mutable_clusters(0), protocol_options);
+        });
+  }
+  if (downstreamProtocol() != Http::CodecClient::Type::HTTP1) {
+    config_helper_.addConfigModifier(
+        [&](envoy::extensions::filters::network::http_connection_manager::v3::HttpConnectionManager&
+                hcm) -> void { hcm.mutable_http2_protocol_options()->set_allow_connect(true); });
+  }
+  config_helper_.addFilter(R"EOF(
+name: envoy.filters.http.compressor
+typed_config:
+  "@type": type.googleapis.com/envoy.extensions.filters.http.compressor.v3.Compressor
+  compressor_library:
+    name: test
+    typed_config:
+      "@type": type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip
+)EOF");
+  HttpProtocolIntegrationTest::initialize();
+}
+
+void WebsocketWithCompressorIntegrationTest::performUpgrade(
+    const Http::TestRequestHeaderMapImpl& upgrade_request_headers,
+    const Http::TestResponseHeaderMapImpl& upgrade_response_headers) {
+  // Establish the initial connection.
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+
+  // Send websocket upgrade request
+  auto encoder_decoder = codec_client_->startRequest(upgrade_request_headers);
+  request_encoder_ = &encoder_decoder.first;
+  response_ = std::move(encoder_decoder.second);
+  test_server_->waitForCounterGe("http.config_test.downstream_cx_upgrades_total", 1);
+  test_server_->waitForGaugeGe("http.config_test.downstream_cx_upgrades_active", 1);
+
+  // Verify the upgrade was received upstream.
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+  validateUpgradeRequestHeaders(upstream_request_->headers(), upgrade_request_headers);
+
+  // Send the upgrade response
+  upstream_request_->encodeHeaders(upgrade_response_headers, false);
+
+  // Verify the upgrade response was received downstream.
+  response_->waitForHeaders();
+  validateUpgradeResponseHeaders(response_->headers(), upgrade_response_headers);
+}
+
+void WebsocketWithCompressorIntegrationTest::sendBidirectionalData() {
+  // Verify that the client can still send data upstream, and that upstream
+  // receives it.
+  codec_client_->sendData(*request_encoder_, "hello", false);
+  ASSERT_TRUE(upstream_request_->waitForData(*dispatcher_, "hello"));
+
+  // Verify the upstream can send data to the client and that the client
+  // receives it.
+  upstream_request_->encodeData("world", false);
+  response_->waitForBodyData(5);
+  EXPECT_EQ("world", response_->body());
+}
+
+// Technically not a websocket tests, but verifies normal upgrades have parity
+// with websocket upgrades
+TEST_P(WebsocketWithCompressorIntegrationTest, NonWebsocketUpgrade) {
```

Comment:
can we also snag

 ProxyingConnectIntegrationTest, ProxyConnect

to make sure CONNECT requests and responses are handled correctly too?